drivers/staging/android/ion/ion_unmaped_heap.c

```diff
@@ -0,0 +1,345 @@
+/*
+ * drivers/staging/android/ion/ion_unmaped_heap.c
+ *
+ * Copyright (C) 2016-2017 Linaro, Inc.
+ * Copyright (C) Allwinner 2014
+ * Author: <sunny@allwinnertech.com> for Allwinner.
+ *
+ * This software is licensed under the terms of the GNU General Public
+ * License version 2, as published by the Free Software Foundation, and
+ * may be copied, distributed, and modified under those terms.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+/*
+ * ION heap type for handling physical memory heap not mapped
+ * in the linux-based OS.
+ *
+ * "unmaped heap" buffers are default not mapped but buffer owner
+ * can explicitly request mapping for some specific purpose.
+ *
+ * Based on Allwinner work (allocation thru gen_pool) and
+ * HiSilicon work (create ION heaps from DT nodes,
+ * Author: Chen Feng <puck.chen@hisilicon.com>).
+ */
+
+#include <linux/err.h>
+#include <linux/errno.h>
+#include <linux/genalloc.h>
+#include <linux/io.h>
+#include <linux/mm.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/platform_device.h>
+#include <linux/scatterlist.h>
+#include <linux/slab.h>
+#include <linux/spinlock.h>
+#include <linux/version.h>
+#include <linux/vmalloc.h>
+
+#include "ion.h"
+#include "ion_priv.h"
+#include "ion_unmaped_heap.h"
+
+/* TODO: could be removed? */
+#define __ALLOCATE_FAIL		-1
+
+struct ion_unmaped_heap {
+	struct ion_heap heap;
+	struct gen_pool *pool;
+	ion_phys_addr_t base;
+	size_t          size;
+};
+
+struct unmaped_buffer_priv {
+	ion_phys_addr_t base;
+};
+
+static ion_phys_addr_t get_buffer_base(struct unmaped_buffer_priv *priv)
+{
+	return priv->base;
+}
+
+static struct device *heap2dev(struct ion_heap *heap)
+{
+	return heap->dev->dev.this_device;
+}
+
+static ion_phys_addr_t ion_unmaped_allocate(struct ion_heap *heap,
+					   unsigned long size,
+					   unsigned long align)
+{
+	struct ion_unmaped_heap *umh =
+		container_of(heap, struct ion_unmaped_heap, heap);
+	unsigned long offset = gen_pool_alloc(umh->pool, size);
+
+	if (!offset) {
+		dev_err(heap2dev(heap),
+			"%s(%d) err: alloc 0x%08x bytes failed\n",
+			__func__, __LINE__, (u32)size);
+		return __ALLOCATE_FAIL;
+	}
+
+	return offset;
+}
+
+static void ion_unmaped_free(struct ion_heap *heap, ion_phys_addr_t addr,
+			    unsigned long size)
+{
+	struct ion_unmaped_heap *umh =
+		container_of(heap, struct ion_unmaped_heap, heap);
+
+	if (addr == __ALLOCATE_FAIL)
+		return;
+	gen_pool_free(umh->pool, addr, size);
+}
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(4,9,0))
```

Comment:
i made this coed to be able to port on various linux kernels since ION heap API recently changed (from 4.8 to 4.9). I you feel this code should be strictly dedicated the linux-swg/linux.git "optee" branch, we can remove useless conditioned code.